crates/apollo-compiler/tests/introspection_split.rs

```diff
@@ -0,0 +1,382 @@
+use apollo_compiler::execution::SchemaIntrospectionSplit;
+use apollo_compiler::ExecutableDocument;
+use apollo_compiler::Schema;
+use expect_test::expect;
+
+const SCHEMA: &str = r#"
+    schema {
+        query: Fibonacci
+        mutation: Fibonacci
+    }
+
+    type Fibonacci {
+        value: Int
+        next(n: Int = 1): Fibonacci
+    }
+
+    input In {
+        list: [Int]
+    }
+
+    directive @someDir(arg: In) on FRAGMENT_DEFINITION
+"#;
+
+#[test]
+fn test_none() {
+    let doc = r#"
+        query TheOperation { value }
+        query OtherOperation { __schema { description } }
+    "#;
+    let expected = expect!["None"];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_only() {
+    let doc = r#"
+        query TheOperation { __schema { description } }
+        query OtherOperation { value }
+    "#;
+    let expected = expect![[r#"
+        Only:
+        query TheOperation {
+          __schema {
+            description
+          }
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_both() {
+    let doc = r#"
+        query TheOperation { value __schema { description } }
+    "#;
+    let expected = expect![[r#"
+        Introspection parts:
+        query TheOperation {
+          __schema {
+            description
+          }
+        }
+
+        Other parts:
+        query TheOperation {
+          value
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_nested() {
+    let doc = r#"
+        query TheOperation { value next { __schema { description } } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
+            locations: [
+                GraphQLLocation {
+                    line: 2,
+                    column: 43,
+                },
+            ],
+            path: [],
+            extensions: {},
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_nested_in_fragment() {
+    let doc = r#"
+        query TheOperation { ...A next { ...A } }
+        fragment A on Fibonacci { value __schema { description } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
+            locations: [
+                GraphQLLocation {
+                    line: 3,
+                    column: 41,
+                },
+            ],
+            path: [],
+            extensions: {},
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_mutation() {
+    let doc = r#"
+        mutation TheOperation { __schema { description } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
```

Comment:
to triple check, we only split on query _operations_? sometimes people use the word "query" to also mean mutations and subscriptions, which would make this error inaccurate.

Reply:
Yes, “query” is used here to mean `OperationType::Query`. Conversely, `__schema` and `__type` are *not* supported:

* Anywhere in a mutation
* Anywhere in a subscription
* In a nested field (not at the root) of a query

Reply:
I rephrased the error messages (see latest commit) to avoid this ambiguity.